File: python/openharness/query.py
"""Streaming ``query()`` entry point.

Spawns the `oh` CLI as a subprocess and yields typed events as they arrive
from ``oh run --output-format stream-json``.
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import os
from collections.abc import AsyncIterator, Awaitable, Callable, Sequence
from typing import Any, Literal

from ._binary import find_oh_binary
from ._permission_server import PermissionCallback
from ._signals import terminate_signal
from ._tools_runtime import prepare_tools_runtime
from .events import Event, parse_event
from .exceptions import OpenHarnessError

__all__ = ["query"]

PermissionMode = Literal["ask", "trust", "deny", "acceptEdits", "plan", "auto", "bypassPermissions"]


def _build_argv(
    oh: str,
    prompt: str,
    *,
    model: str | None,
    permission_mode: PermissionMode,
    allowed_tools: Sequence[str] | None,
    disallowed_tools: Sequence[str] | None,
    max_turns: int,
    system_prompt: str | None,
) -> list[str]:
    """Assemble the argv for ``oh run ... --output-format stream-json``."""
    argv: list[str] = [oh, "run", prompt, "--output-format", "stream-json"]
    if model:
        argv += ["--model", model]
    if permission_mode:
        argv += ["--permission-mode", permission_mode]
    if allowed_tools:
        argv += ["--allowed-tools", ",".join(allowed_tools)]
    if disallowed_tools:
        argv += ["--disallowed-tools", ",".join(disallowed_tools)]
    if max_turns is not None:
        argv += ["--max-turns", str(max_turns)]
    if system_prompt:
        argv += ["--system-prompt", system_prompt]
    return argv


async def query(
    prompt: str,
    *,
    model: str | None = None,
    permission_mode: PermissionMode = "trust",
    allowed_tools: Sequence[str] | None = None,
    disallowed_tools: Sequence[str] | None = None,
    max_turns: int = 20,
    system_prompt: str | None = None,
    cwd: str | os.PathLike[str] | None = None,
    env: dict[str, str] | None = None,
    tools: Sequence[Callable[..., Any] | Callable[..., Awaitable[Any]]] | None = None,
    can_use_tool: PermissionCallback | None = None,
) -> AsyncIterator[Event]:
    """Run a single prompt through openHarness and stream events.

    Example::

        async for event in query("What is 2+2?", model="ollama/llama3", max_turns=1):
            print(event)

    :param prompt: The user prompt to send.
    :param model: Model string (e.g. ``"ollama/llama3"``, ``"claude-sonnet-4-6"``).
        When omitted, ``oh`` uses the model from ``.oh/config.yaml`` or env detection.
    :param permission_mode: One of the CLI-supported modes; ``"trust"`` by default so
        headless scripts don't hang on permission prompts.
    :param allowed_tools: Whitelist of tool names. When set, tools outside this list
        are blocked.
    :param disallowed_tools: Blacklist of tool names.
    :param max_turns: Maximum number of model turns before the session ends.
    :param system_prompt: Override the default system prompt.
    :param cwd: Working directory for the spawned CLI.
    :param env: Environment variables to pass to the CLI. Merged on top of
        ``os.environ``.
    :param tools: Python callables to expose as MCP tools for this query.
        When set, an in-process MCP HTTP server is started and injected
        into an ephemeral ``.oh/config.yaml``; the subprocess runs with
        that temp dir as its cwd. Any existing user config at ``cwd`` is
        preserved (copied over, then the server entry is appended).
    :param can_use_tool: Optional sync or async callback invoked on every
        permission check. Receives a context dict (``toolName``,
        ``toolInputJson``, etc.) and must return ``"allow"``,
        ``"deny"``, or ``"ask"`` — optionally wrapped as
        ``{"decision": "...", "reason": "..."}``. Mirrors Claude Code's
        ``can_use_tool`` SDK hook. When set, the SDK starts an
        in-process HTTP server and wires a ``permissionRequest`` hook in
        the ephemeral config.

    :yields: Typed :class:`Event` objects matching NDJSON lines emitted by
        the CLI.

    :raises OhBinaryNotFoundError: if the ``oh`` CLI can't be located.
    :raises OpenHarnessError: if the subprocess exits non-zero.
    """
    oh = find_oh_binary()
    argv = _build_argv(
        oh,
        prompt,
        model=model,
        permission_mode=permission_mode,
        allowed_tools=allowed_tools,
        disallowed_tools=disallowed_tools,
        max_turns=max_turns,
        system_prompt=system_prompt,
    )

    merged_env = {**os.environ, **(env or {})}

    runtime = None
    effective_cwd: str | os.PathLike[str] | None = cwd
    if tools or can_use_tool is not None:
        runtime = await prepare_tools_runtime(
            tools,
            base_cwd=str(cwd) if cwd else None,
            can_use_tool=can_use_tool,
        )
        effective_cwd = runtime.cwd

    proc = await asyncio.create_subprocess_exec(
        *argv,
        stdin=asyncio.subprocess.DEVNULL,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        cwd=effective_cwd,
        env=merged_env,
    )

    assert proc.stdout is not None
    try:
        async for raw_line in proc.stdout:
            line = raw_line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                # Non-JSON line (e.g. transient warning) — skip.
                continue
            yield parse_event(obj)
    except asyncio.CancelledError:
        if proc.returncode is None:
            with contextlib.suppress(ProcessLookupError, OSError):
                proc.send_signal(terminate_signal())
            try:
                await asyncio.wait_for(proc.wait(), timeout=5)
            except TimeoutError:
                proc.kill()
        raise
    finally:
        rc = await proc.wait()
        if runtime is not None:
            await runtime.close()
        if rc != 0:
            stderr = ""
            if proc.stderr is not None:
                with contextlib.suppress(Exception):
                    stderr_bytes = await proc.stderr.read()
                    stderr = stderr_bytes.decode(errors="replace")
            raise OpenHarnessError(
                f"'oh run' exited with code {rc}",
                stderr=stderr,
                exit_code=rc,
            )

File: packages/sdk/src/query.ts
/**
 * Streaming `query()` — single-shot prompt against a fresh `oh` subprocess.
 * Mirror of `python/openharness/query.py`.
 */

import { spawn } from "node:child_process";
import process from "node:process";
import { OpenHarnessError } from "./errors.js";
import { type Event, parseEvent } from "./events.js";
import { findOhBinary } from "./internal/binary.js";
import { splitNdjson } from "./internal/ndjson.js";
import { sendKill, sendTerminate } from "./internal/signals.js";
import { prepareToolsRuntime, type ToolsRuntime } from "./internal/tools-runtime.js";
import type { OpenHarnessOptions } from "./options.js";

const DEFAULT_MAX_TURNS = 20;
const DEFAULT_PERMISSION_MODE = "trust" as const;
const KILL_TIMEOUT_MS = 5_000;

export function buildArgv(prompt: string, options: OpenHarnessOptions = {}): string[] {
  const argv: string[] = ["run", prompt, "--output-format", "stream-json"];
  if (options.model) argv.push("--model", options.model);
  argv.push("--permission-mode", options.permissionMode ?? DEFAULT_PERMISSION_MODE);
  if (options.allowedTools && options.allowedTools.length > 0) {
    argv.push("--allowed-tools", options.allowedTools.join(","));
  }
  if (options.disallowedTools && options.disallowedTools.length > 0) {
    argv.push("--disallowed-tools", options.disallowedTools.join(","));
  }
  argv.push("--max-turns", String(options.maxTurns ?? DEFAULT_MAX_TURNS));
  if (options.systemPrompt) argv.push("--system-prompt", options.systemPrompt);
  return argv;
}

/**
 * Run a single prompt and stream typed events as they arrive.
 *
 * @example
 * ```ts
 * for await (const event of query("What is 2+2?", { model: "ollama/llama3" })) {
 *   if (event.type === "text") process.stdout.write(event.content);
 * }
 * ```
 *
 * @throws {OhBinaryNotFoundError} if the `oh` CLI cannot be located.
 * @throws {OpenHarnessError} if the subprocess exits non-zero.
 */
export async function* query(prompt: string, options: OpenHarnessOptions = {}): AsyncGenerator<Event, void, void> {
  const handle = findOhBinary(options.ohBinary);
  const argv = [...handle.prefixArgs, ...buildArgv(prompt, options)];
  const env = { ...process.env, ...(options.env ?? {}) };

  let runtime: ToolsRuntime | null = null;
  let effectiveCwd = options.cwd;
  if (options.tools && options.tools.length > 0) {
    runtime = await prepareToolsRuntime({ tools: options.tools, baseCwd: options.cwd });
    effectiveCwd = runtime.cwd;
  }

  const proc = spawn(handle.command, argv, {
    stdio: ["ignore", "pipe", "pipe"],
    cwd: effectiveCwd,
    env,
    windowsHide: true,
  });

  let stderrBuf = "";
  proc.stderr?.setEncoding("utf8");
  proc.stderr?.on("data", (chunk: string) => {
    stderrBuf += chunk;
    if (stderrBuf.length > 64 * 1024) {
      // Cap retained stderr to prevent unbounded growth on a noisy CLI run.
      stderrBuf = stderrBuf.slice(-64 * 1024);
    }
  });

  const exitPromise = new Promise<{ code: number | null; signal: NodeJS.Signals | null }>((resolve) => {
    proc.once("exit", (code, signal) => resolve({ code, signal }));
  });

  let abnormalExit = true;
  try {
    if (!proc.stdout) {
      throw new OpenHarnessError("subprocess stdout was not piped");
    }
    for await (const obj of splitNdjson(proc.stdout)) {
      yield parseEvent(obj);
    }
    abnormalExit = false;
  } finally {
    if (abnormalExit && proc.exitCode == null && proc.signalCode == null) {
      sendTerminate(proc);
      const timer = setTimeout(() => sendKill(proc), KILL_TIMEOUT_MS);
      try {
        await exitPromise;
      } finally {
        clearTimeout(timer);
      }
    } else {
      await exitPromise;
    }
    if (runtime) await runtime.close();
  }

  // Only surface a non-zero exit when the stream completed naturally — if the
  // consumer broke out or the body threw, we already initiated termination
  // and adding another error on top would mask the original cause.
  if (!abnormalExit) {
    const { code } = await exitPromise;
    if (code !== 0 && code !== null) {
      throw new OpenHarnessError(`'oh run' exited with code ${code}`, {
        stderr: stderrBuf,
        exitCode: code,
      });
    }
  }
}
